File: Data Structures & Algorithms/binary-tree-preorder-traversal/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        res = []
        def preOrder(curr):
            if not curr:
                return
            res.append(curr.val)
            preOrder(curr.left)
            preOrder(curr.right)
        preOrder(root)
        return res

File: Data Structures & Algorithms/closest-binary-search-tree-value/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int closestValue(TreeNode root, double target) {
        List<Integer> nums = new ArrayList<>();
        inorder(root, nums);
        return Collections.min(nums, new Comparator<Integer>() {
            public int compare(Integer o1, Integer o2) {
                return Math.abs(o1 - target) < Math.abs(o2 - target) ? -1: 1;
            }
        });
    }
    void inorder(TreeNode curr,List<Integer> nums) {
        if (curr == null) {
            return;
        }
        inorder(curr.left, nums);
        nums.add(curr.val);
        inorder(curr.right, nums);
    }
}
